File: python3/Hash Maps and Sets/geometric_sequence_triplets.py
from collections import defaultdict
from typing import List


def geometric_sequence_triplets(nums: List[int], r: int) -> int:
    # Use 'defaultdict' to ensure the default value of 0 is returned when 
    # accessing a key that doesn’t exist in the hash map. This effectively sets 
    # the default frequency of all elements to 0.
    left_map = defaultdict(int)
    right_map = defaultdict(int)
    count = 0
    # Populate 'right_map' with the frequency of each element in the array.
    for x in nums:
        right_map[x] += 1
    # Search for geometric triplets that have x as the center.
    for x in nums:
        # Decrement the frequency of x in 'right_map' since x is now being
        # processed and is no longer to the right.
        right_map[x] -= 1
        if x % r == 0:
            count += left_map[x // r] * right_map[x * r]
        # Increment the frequency of x in 'left_map' since it'll be a part of the
        # left side of the array once we iterate to the next value of x.
        left_map[x] += 1
    return count

File: typescript/Hash Maps and Sets/geometric_sequence_triplets.ts
function geometricSequenceTriplets(nums: number[], r: number): number {
    const leftMap: { [key: number]: number } = {};
    const rightMap: { [key: number]: number } = {};
    let count = 0;
    // Populate 'rightMap' with the frequency of each element in the array.
    for (const x of nums)
        rightMap[x] = (rightMap[x] ?? 0) + 1;
    // Search for geometric triplets that have x as the center.
    for (const x of nums) {
        // Decrement the frequency of x in rightMap since x is now being
        // processed and is no longer to the right.
        rightMap[x] -= 1;
        if (x % r === 0)
            count += (leftMap[x / r] ?? 0) * (rightMap[x * r] ?? 0);
        // Increment the frequency of x in leftMap since it'll be a part of the
        // left side of the array once we iterate to the next value of x.
        leftMap[x] = (leftMap[x] ?? 0) + 1;
    }
    return count;
}